src/client/sound_openal_internal.cpp

```diff
@@ -0,0 +1,1146 @@
+		warningstream << "Audio: OpenAL error: " << getAlErrorString(error)
+				<< "preparing sound buffer for sound \""
+				<< decode_info.name_for_logging << "\"" << std::endl;
+	}
+
+	return snd_buffer_id;
+}
+
+/*
+ * SoundManagerSingleton class
+ */
+
+bool SoundManagerSingleton::init()
+{
+	if (!(m_device = unique_ptr_alcdevice(alcOpenDevice(nullptr)))) {
+		errorstream << "Audio: Global Initialization: Failed to open device" << std::endl;
+		return false;
+	}
+
+	if (!(m_context = unique_ptr_alccontext(alcCreateContext(m_device.get(), nullptr)))) {
+		errorstream << "Audio: Global Initialization: Failed to create context" << std::endl;
+		return false;
+	}
+
+	if (!alcMakeContextCurrent(m_context.get())) {
+		errorstream << "Audio: Global Initialization: Failed to make current context" << std::endl;
+		return false;
+	}
+
+	alDistanceModel(AL_INVERSE_DISTANCE_CLAMPED);
+
+	// Speed of sound in nodes per second
+	// FIXME: This value assumes 1 node sidelength = 1 meter, and "normal" air.
+	//        Ideally this should be mod-controlled.
+	alSpeedOfSound(343.3f);
+
+	readDopplerFactorFromSettings("", nullptr);
+	g_settings->registerChangedCallback("sound_doppler_factor",
+			readDopplerFactorFromSettings, nullptr);
+
+	if (alGetError() != AL_NO_ERROR) {
+		errorstream << "Audio: Global Initialization: OpenAL Error " << alGetError() << std::endl;
+		return false;
+	}
+
+	infostream << "Audio: Global Initialized: OpenAL " << alGetString(AL_VERSION)
+		<< ", using " << alcGetString(m_device.get(), ALC_DEVICE_SPECIFIER)
+		<< std::endl;
+
+	return true;
+}
+
+SoundManagerSingleton::~SoundManagerSingleton()
+{
+	g_settings->deregisterChangedCallback("sound_doppler_factor",
+			readDopplerFactorFromSettings, nullptr);
+
+	infostream << "Audio: Global Deinitialized." << std::endl;
+}
+
+void SoundManagerSingleton::readDopplerFactorFromSettings(const std::string &, void *)
+{
+	f32 doppler_factor = MYMAX(g_settings->getFloat("sound_doppler_factor"), 0.0f); // 0.0f if nan
+	alDopplerFactor(doppler_factor);
+}
+
+/*
+ * ISoundDataOpen struct
+ */
+
+std::shared_ptr<ISoundDataOpen> ISoundDataOpen::fromOggFile(std::unique_ptr<RAIIOggFile> oggfile,
+		const std::string &filename_for_logging)
+{
+	// Get some information about the OGG file
+	Optional<OggFileDecodeInfo> decode_info = oggfile->getDecodeInfo(filename_for_logging);
+	if (!decode_info.has_value()) {
+		warningstream << "Audio: Error decoding "
+				<< filename_for_logging << std::endl;
+		return nullptr;
+	}
+
+	// use duration (in seconds) to decide whether to load all at once or to stream
+	if (decode_info->length_seconds <= SOUND_DURATION_MAX_SINGLE) {
+		return std::make_shared<SoundDataOpenSinglebuf>(std::move(oggfile), *decode_info);
+	} else {
+		return std::make_shared<SoundDataOpenStream>(std::move(oggfile), *decode_info);
+	}
+}
+
+/*
+ * SoundDataUnopenBuffer struct
+ */
+
+std::shared_ptr<ISoundDataOpen> SoundDataUnopenBuffer::open(const std::string &sound_name) &&
```

Comment:
I don't get what "unopen" means here

Reply:
Added comment to SoundDataUnopenBuffer struct. (See also comment below.)